### PR TITLE
add Parent for pom model manipulation

### PR DESCRIPTION
**Pull request by holiday-sunrise**

### CODE REVIEW

**Comment by holiday-sunrise** [comment]
can you merg it ?

**Comment by holiday-sunrise** [comment]
What is the problem with PR.  have you ever worked in a compony with 24 Business groups and 300 API`s.

**Comment by afelisatti** [comment]
Hi, @holiday-sunrise! Thanks for this contribution. As mentioned on the backing issue (#466), we were confused by the lack of usage of the new entities and wanted to better understand the intent behind them. We are looking into all PRs and hope you'll be patient as we work on getting them into the codebase.

**Comment by holiday-sunrise** [comment]
Hello Ane I would like to try to explain it to you better. We at OBI have 300 API's that are in production. The C4E ensures that currently 24 business groups can migrate their API's automatically in a build process.  No developer has to migrate his API locally, it happens through a build pipeline.

Exactly for this automatic process we have the need to change the parent in an API. 

In a Mule3 API different settings have to be set company wide than in a Mule4 API. Other default connectors, other unified excption handler, common dataweaves libraries must be used to change the settings to our artifactory. As this is set a Maven parent for the API.

Now the MMA comes into play without the PR the Maven parent of an API cannot be changed. So we added the changes for the PomModel and the Parent. 

Example from a Mule 3 API

```
	<parent>
		<groupId>com.obi.smartgate.maven.parent</groupId>
		<artifactId>smartgate-parent-api-rest</artifactId>
		<version>1.8.0</version>
	</parent>
```

in a migrated Mule4 API it then looks like this

```
  <parent>
    <groupId>com.obi.smartgate.maven.parent</groupId>
    <artifactId>smartgate-parent-mule-api</artifactId>
    <version>2.1.0</version>
  </parent>
```

I hope it explains why we need the extension. 

kind reagards

rainer

--------------------------------------------------------------------------------------------------------------------------

We have written 37 task/steps in our fork at OBI that fix bugs in the MMA or are enhancements.  We would like to bring more into the community but it doesn't seem that tha MMA is a community project from our point of view. Because not the community decides what goes into the MMA but the Mule product management. Think about it

**Comment by holiday-sunrise** [comment]
image

Parent has now 100 unit coverage

**Comment by afelisatti** [comment]
Thanks, @holiday-sunrise! @mlischetti will review the changes soon.

File: mule-migration-tool-api/src/main/java/com/mulesoft/tools/migration/project/model/pom/PomModel.java [hunk 1]
```diff
@@ -156,6 +156,24 @@ public void setArtifactId(String artifactId) {
...
+   * @return a {@link String}
+   */
+  public Parent getParent() {
+    return new Parent(model.getParent());
+  }
+
+  /**
+   * Sets the artifact id in the pom.
+   *
+   * @param artifactId
```
Comment (mlischetti) [inline mule-migration-tool-api/src/main/java/com/mulesoft/tools/migration/project/model/pom/PomModel.java:171]: change _artifactId_ by _parent_

File: mule-migration-tool-api/src/main/java/com/mulesoft/tools/migration/project/model/pom/Parent.java [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+ * 
+ * @author Mulesoft Inc.
+ * @since 1.0.0
+ */
+public class Parent {
+
+  private final org.apache.maven.model.Parent parent;
+
+  protected Parent(org.apache.maven.model.Parent parent) {
+    this.parent = parent;
```
Comment (mlischetti) [inline mule-migration-tool-api/src/main/java/com/mulesoft/tools/migration/project/model/pom/Parent.java:19]: Add validation for a null value parent. You can use guava preconditions:
`Preconditions.checkArgument(parent != null,"The parent cannot be null");`

File: mule-migration-tool-api/src/main/java/com/mulesoft/tools/migration/project/model/pom/Parent.java [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+    return parent.getRelativePath();
+  }
+
+  /**
+   * Sets the parent artifact id.
+   *
+   * @param artifactId
+   */
+  public void setArtifactId(String artifactId) {
+    if (parent != null)
```
Comment (mlischetti) [inline mule-migration-tool-api/src/main/java/com/mulesoft/tools/migration/project/model/pom/Parent.java:77]: In which scenario parent attribute could be null? (If It's not really necessary please remove it)

File: mule-migration-tool-api/src/main/java/com/mulesoft/tools/migration/project/model/pom/Parent.java [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+      parent.setArtifactId(artifactId);
+  }
+
+  /**
+   * Sets the parent group id.
+   *
+   * @param groupId
+   */
+  public void setGroupId(String groupId) {
+    if (parent != null)
```
Comment (mlischetti) [inline mule-migration-tool-api/src/main/java/com/mulesoft/tools/migration/project/model/pom/Parent.java:87]: Same above comment

File: mule-migration-tool-api/src/main/java/com/mulesoft/tools/migration/project/model/pom/Parent.java [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+      parent.setGroupId(groupId);
+  }
+
+  /**
+   * Sets the parent version.
+   *
+   * @param version
+   */
+  public void setVersion(String version) {
+    if (parent != null)
```
Comment (mlischetti) [inline mule-migration-tool-api/src/main/java/com/mulesoft/tools/migration/project/model/pom/Parent.java:97]: Idem

File: mule-migration-tool-api/src/main/java/com/mulesoft/tools/migration/project/model/pom/Parent.java [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+      parent.setVersion(version);
+  }
+
+  /**
+   * Retrieves the parent relativePath.
+   *
+   * @return a {@link String}
+   */
+  public void setRelativePath(String relativePath) {
+    if (parent != null)
```
Comment (mlischetti) [inline mule-migration-tool-api/src/main/java/com/mulesoft/tools/migration/project/model/pom/Parent.java:107]: Idem

File: mule-migration-tool-api/src/main/java/com/mulesoft/tools/migration/project/model/pom/Parent.java [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+      parent.setRelativePath(relativePath);
+  }
+}
```
Comment (mlischetti) [inline mule-migration-tool-api/src/main/java/com/mulesoft/tools/migration/project/model/pom/Parent.java:110]: Add a parent builder class to build Parent instances (e.g ParentBuilder class). You can use the same pattern as PomModelBuilder, DependencyBuilder, RepositoryBuilder, etc.

File: mule-migration-tool-api/src/test/java/com/mulesoft/tools/migration/project/model/pom/PomModelTestCase.java [hunk 8]
```diff
@@ -225,5 +266,36 @@ public void setDescription() {
...
+    public void setParentRelativePath() {
+      model.getParent().setRelativePath("new-path");
+      assertThat("RelativePath is not the expected", model.getParent().getRelativePath(), equalTo("new-path"));
+    }
+
+    @Test
+    public void setParent() {
+      Parent parent = new Parent();
+      model.setParent(parent);
+      assertNotNull(model.getParent());
```
Comment (mlischetti) [inline mule-migration-tool-api/src/test/java/com/mulesoft/tools/migration/project/model/pom/PomModelTestCase.java:298]: Assert that parent it's actually modified. e.g
```
@Test
public void setParent() {
  Parent parent = new Parent();
  parent.setGroupId("myGroupId");
  parent.setArtifactId("myArtifactId");
  parent.setVersion("1.0.0");
  model.setParent(parent);
  assertNotNull(model.getParent());
  assertThat("Parent groupId is not the expected", model.getParent().getGroupId(), equalTo("myGroupId"));
  assertThat("Parent artifactId is not the expected", model.getParent().getArtifactId(), equalTo("myArtifactId"));
  assertThat("Parent version is not the expected", model.getParent().getVersion(), equalTo("1.0.0"));
}
```
Using a parent builder class this construction can be simplified
```
@Test
public void setParent() {
  Parent parent =
      new Parent.ParentBuilder().withGroupId("myGroupId").withArtifactId("myArtifactId").withVersion("1.0.0").build();
  model.setParent(parent);
  assertNotNull(model.getParent());
  assertThat("Parent groupId is not the expected", model.getParent().getGroupId(), equalTo("myGroupId"));
  assertThat("Parent artifactId is not the expected", model.getParent().getArtifactId(), equalTo("myArtifactId"));
  assertThat("Parent version is not the expected", model.getParent().getVersion(), equalTo("1.0.0"));
}
```

File: mule-migration-tool-api/src/main/java/com/mulesoft/tools/migration/project/model/pom/Parent.java (hunk 6 above)
Comment (holiday-sunrise) [reply mule-migration-tool-api/src/main/java/com/mulesoft/tools/migration/project/model/pom/Parent.java:107]: removed

File: mule-migration-tool-api/src/main/java/com/mulesoft/tools/migration/project/model/pom/Parent.java (hunk 5 above)
Comment (holiday-sunrise) [reply mule-migration-tool-api/src/main/java/com/mulesoft/tools/migration/project/model/pom/Parent.java:97]: removed

File: mule-migration-tool-api/src/main/java/com/mulesoft/tools/migration/project/model/pom/Parent.java (hunk 4 above)
Comment (holiday-sunrise) [reply mule-migration-tool-api/src/main/java/com/mulesoft/tools/migration/project/model/pom/Parent.java:87]: removed

File: mule-migration-tool-api/src/main/java/com/mulesoft/tools/migration/project/model/pom/Parent.java (hunk 7 above)
Comment (holiday-sunrise) [reply mule-migration-tool-api/src/main/java/com/mulesoft/tools/migration/project/model/pom/Parent.java:110]: A builder for a wrapper is not needed !?! 

Why create Builder with 100 lines of. The code have also to be tested.

Keep it easy and write code that is really needed.

File: mule-migration-tool-api/src/main/java/com/mulesoft/tools/migration/project/model/pom/Parent.java (hunk 2 above)
Comment (holiday-sunrise) [reply mule-migration-tool-api/src/main/java/com/mulesoft/tools/migration/project/model/pom/Parent.java:19]: the constructor is protected and there is only one caller. Not needed

File: mule-migration-tool-api/src/test/java/com/mulesoft/tools/migration/project/model/pom/PomModelTestCase.java (hunk 8 above)
Comment (holiday-sunrise) [reply mule-migration-tool-api/src/test/java/com/mulesoft/tools/migration/project/model/pom/PomModelTestCase.java:298]: A builder for a wrapper is not needed !?! 

Why create Builder with 100 lines of. The code have also to be tested.

Keep it easy and write code that is really needed.

**Comment by holiday-sunrise** [comment]
what can i do ?

**Comment by mlischetti** [comment]
You can change your test by something like this
```
    @Test
    public void setParent() {
      Parent parent =
          new Parent.ParentBuilder().withGroupId("myGroupId").withArtifactId("myArtifactId").withVersion("1.0.0").build();
      assertNotNull(model.getParent());
      model.setParent(parent);
      assertNotNull(model.getParent());
      assertThat("Parent is not the expected", parent.getInnerModel(), equalTo(model.getParent().getInnerModel()));
    }
```

**Review by mlischetti — APPROVED** [review]

**Review by mlischetti — APPROVED** [review]
LGMT

**Pull request merged**